Add and remove peaks in new Instrument View

jclarkeSTFC opened the pull request:
Add and remove peaks in the new Instrument View. If a single detector is selected then one can now select `Add Peak` and select the peak location in the line plot. If a single peaks workspace is already selected to overlay then the peak will be added to that workspace, otherwise it will be added to a separate Instrument View peaks workspace. When one or more detectors are selected, one can select `Delete Single Peak` and click in the line plot, then the peak closest to where you clicked will be removed from its peaks workspace. The final option is to delete all peaks on all selected detectors.

Closes #40045 . <!-- One line per closed issue. -->

<!-- If issue raised by user. Do not leak email addresses.
**Report to:** [user name]
-->

### To test:

Add peaks, delete peaks (including in the old Instrument View, I suggest having both open), do some masking at the same time. Use the old Instrument View to check that peaks have been added and removed correctly.

<!-- Include sufficient instructions for someone unfamiliar with the application to test.
Ok to refer back to instructions in the issue.
-->

<!-- REMEMBER:
- Add labels, milestones, etc.
- Ensure the base of this PR is correct (e.g. release-next or main)
- Add release notes in separate file as per ([guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)), or justify their absence:
  *This does not require release notes* because <fill in an explanation of why>
-->

<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?
- If the PR author isn’t in the `mantid-developers` or `mantid-contributors` teams, add a review comment `rerun ci` to authorize/rerun the CI

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

jclarkeSTFC committed:
Add peak adding functionality

If one peaks workspace is selected in the list then add peaks to it,
otherwise add peaks in a special workspace like the old instrument
view.

jclarkeSTFC committed:
Add delete single peak functionality

This will delete the peak nearest to the place where the user clicks in
the line plot.

jclarkeSTFC committed:
Add button to delete peaks on selected detectors

Every peak on any selected detector will be deleted from the respective
peaks workspace. Exposed removePeaks() to Python so I could remove
multiple peaks at once. This makes it easier because otherwise after
every peak removal I would have to recalculate the peak indices, plus
the interface would be refreshed on the peaks workspace changing.

jclarkeSTFC committed:
Add tests for new adding/deleting peaks functionality

jclarkeSTFC committed:
Suppress error in case of no connection

jclarkeSTFC committed:
Fix ADS callback deadlock when adding peaks

Calling CreatePeaksWorkspace without StoreInADS=False will result
the add callback on the ADS observer being called before
CreatePeaksWorkspace has exited. This will end up calling a different
algorithm which cannot proceed because CreatePeaksWorkspace is still
running, hence a deadlock. The fix is to add the peaks workspace
manually, then CreatePeaksWorkspace won't be running and other
algorithms will be free to run.

jclarkeSTFC committed:
Add release note

RichardWaiteSTFC commented on qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py:
```diff
@@ -383,9 +388,64 @@ def peak_overlay_points(self) -> list[list[DetectorPeaks]]:
+            if peaks_ws.name() not in peaks_grouped_by_ws:
+                continue
+            peaks_by_detector = peaks_grouped_by_ws[peaks_ws.name()]
+            picked_detector_peaks = [p for p in peaks_by_detector if p.detector_id in detector_ids]
```
The peaks workspace has a `DetID` column which means this search for matching detector indices could be vectorised? It might not make much difference!

RichardWaiteSTFC commented on qt/python/instrumentview/instrumentview/FullInstrumentViewWindow.py:
```diff
@@ -572,6 +601,9 @@ def set_plotter_scalar_bar_range(self, clim: tuple[int, int], label: str) -> Non
     def closeEvent(self, QCloseEvent: QEvent) -> None:
         """When closing, make sure to close the plotters and figure correctly to prevent errors"""
         super().closeEvent(QCloseEvent)
+        with suppress(TypeError):
```
This looks like it fixes an unrelated bug - does it require a separate release note?

RichardWaiteSTFC reviewed (changes requested):
Thanks for this - I've had a look through but not tested yet sorry!

jclarkeSTFC committed:
Merge branch 'main' into 40045-add-remove-peaks

jclarkeSTFC committed:
Use diffractometer constants for unit conversion

These are updated by the calibration, and will give different numbers
to the values from the detector info.

Co-authored-by: RichardWaiteSTFC

jclarkeSTFC committed:
Use ADS.doesExist instead of loop

Co-authored-by: RichardWaiteSTFC

jclarkeSTFC committed:
Update year in copyright notices

Co-authored-by: <RichardWaiteSTFC>

RichardWaiteSTFC commented on qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py:
```diff
@@ -56,6 +59,8 @@ def setup(self):
         self._workspace_x_unit = x_unit.unitID()
         self._workspace_x_unit_display = f"{str(x_unit.caption())} ({str(x_unit.symbol())})"
         self._selected_peaks_workspaces = []
+        self._instrument_view_peaks_ws_name = f"instrument_view_peaks_{self._workspace.name()}"
```
Is there any reason to preserve backward compatibility by naming the workspace `SingleCrystalPeakTable`?
I do like however the table doesn't overwrite a table produced on another workspace.

jclarkeSTFC replied:
I don't know of a specific reason

RichardWaiteSTFC reviewed (changes requested):
Sorry, small bug with unit conversion still!

Also a usability thing - you often need to zoom into a peak on 1D plot before adding. Here you have to remember to do that before clicking Add.
If you do zoom into the 1d plot but leave the zoom button clicked and add a peak the red line doesn't show up in the plot - is that easy to fix?

jclarkeSTFC committed:
Include l2 in unit conversion

jclarkeSTFC committed:
Disable line plot toolbar when adding a peak

Trying to zoom or do anything else is not possible because the
interface is waiting for the peak to be picked.

jclarkeSTFC committed:
Fix problem picking a detector while adding peak

If the user is in the middle of adding a peak, then they select a
detector, then stop the adding peak process. You can only add a peak
if exactly one detector is selected, otherwise we don't know which
detector the peak is on.

GuiMacielPereira reviewed (changes requested):
Everything is working very well, I like that it's very intuitive to add/remove peaks, the peaks from the old and new instrument view are stored into separate workspaces with different colors.

I picked up two issues, the first one is that when I click to delete all peaks I get the error
```
Traceback (most recent call last):
  File "/home/gui/mantid/qt/python/instrumentview/instrumentview/FullInstrumentViewPresenter.py", line 485, in on_delete_all_selected_peaks_clicked
    self._model.delete_peaks_on_all_selected_detectors()
  File "/home/gui/mantid/qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py", line 447, in delete_peaks_on_all_selected_detectors
    peaks_ws.removePeaks(peaks_to_remove)
    ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'PeaksWorkspace' object has no attribute 'removePeaks'
```

The second is that the old and new instrument view synchronise well, but when I delete a peak in the old one, it's not deleted automatically in the new one (when I add a peak in the old one, the peak appears immediately in the new one).

RichardWaiteSTFC reviewed (approved):
Thanks for fixing unit conversion. In my testing it does indeed add the peak at the right place even though the red line isn't plotted when the zoom button is selected.

I'm happy to approve - but if you did want to revisit this in another PR my suggestion would be to allow the user to select zoom tool while adding a peak, but don't plot red line or allow user to add peak whilst zoom is enabled. This way the user can fiddle with zoom until can add peak centre with accuracy, then disable zoom and add a peak (without having to toggle add peak button). But that is just my opinion, it feels like more button presses required to add peaks relative to old instrument view. But that's just my opinion - feel free to ask WISH!

RichardWaiteSTFC commented:
> Everything is working very well, I like that it's very intuitive to add/remove peaks, the peaks from the old and new instrument view are stored into separate workspaces with different colors.
> 
> I picked up two issues, the first one is that when I click to delete all peaks I get the error
> 
> ```
> Traceback (most recent call last):
>   File "/home/gui/mantid/qt/python/instrumentview/instrumentview/FullInstrumentViewPresenter.py", line 485, in on_delete_all_selected_peaks_clicked
>     self._model.delete_peaks_on_all_selected_detectors()
>   File "/home/gui/mantid/qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py", line 447, in delete_peaks_on_all_selected_detectors
>     peaks_ws.removePeaks(peaks_to_remove)
>     ^^^^^^^^^^^^^^^^^^^^
> AttributeError: 'PeaksWorkspace' object has no attribute 'removePeaks'
> ```
> 

This worked for me - did you build after changing branch as this PR includes an additional cpp method I think? Otherwise maybe I'm not doing the same as you?

> The second is that the old and new instrument view synchronise well, but when I delete a peak in the old one, it's not deleted automatically in the new one (when I add a peak in the old one, the peak appears immediately in the new one).

I think the ADS observer is working in the new one though, if you open a peak table and delete a row it does update new instrument view (at least for me!).

GuiMacielPereira reviewed (approved):
That was my bad, it's all working well and there is extensive unit testing :+1:
